### PR TITLE
Receive images

### PR DESCRIPTION
**Pull request by penguin42**
This is at the 'seems to work' level; I get the feeling it could probably do with a bit more testing and perhaps a bit more thought on things like what gets free'd where; but have a look.

### CODE REVIEW

File: matrix-api.c [hunk 1]
```diff
@@ -105,6 +105,8 @@ typedef struct {
     gchar *content_type;
     gboolean got_headers;
     JsonParser *json_parser;
+    const char *body;
```
Comment (richvdh) [inline matrix-api.c]: need to initialise these in _response_parser_data_new

File: matrix-room.c [hunk 2]
```diff
@@ -444,6 +450,17 @@ static const char *type_guess(PurpleStoredImage *image)
     }
 }
 
+/**
+ * Check if the declared image type is an image type we recongise.
```
Comment (richvdh) [inline matrix-room.c]: s/image type/content-type/

File: matrix-room.c [hunk 3]
```diff
@@ -444,6 +450,17 @@ static const char *type_guess(PurpleStoredImage *image)
...
+ */
+static gboolean check_image_type(const char *content_type)
```
Comment (richvdh) [inline matrix-room.c]: nit: i'd prefer a less ambiguous name here: `is_image_type` or something

File: matrix-room.c [hunk 4]
```diff
@@ -503,6 +520,142 @@ static void _send_image_hook(MatrixRoomEvent *event, gboolean just_free)
...
+    struct ReceiveImageData *rid = user_data;
+    if (check_image_type(content_type)) {
+        /* Excellent - something to work with */
+        int img_id = purple_imgstore_add_with_id(g_memdup(raw_body, raw_body_len),
+                                                 raw_body_len, NULL);
+        serv_got_chat_in(rid->conv->account->gc, g_str_hash(rid->room_id), rid->sender_display_name,
+                PURPLE_MESSAGE_RECV | PURPLE_MESSAGE_IMAGES,
+                g_strdup_printf("<IMG ID=\"%d\">", img_id), rid->timestamp / 1000);
+    } else {
+        serv_got_chat_in(rid->conv->account->gc, g_str_hash(rid->room_id),
```
Comment (richvdh) [inline matrix-room.c]: it would be nice to show the fallback (event.body) here?

ditto in _image_download_bad_response and _image_download_error

Comment (penguin42) [inline matrix-room.c]: OK, I think I've fixed that and the other comments; and pushed to this same branch - do you see that?

**Comment by richvdh** [comment]
Sorry for sitting on this. LGTM!

**Comment by richvdh** [comment]
(and thanks!)

**Pull request merged**